File: sample/ios:objective-c/HybridTestObjc/HybridTestObjc/ViewController.m
//
//  ViewController.m
//  HybridTestObjc
//
//  Created by Jimmy.강세훈 on 2023/03/21.
//

#import "ViewController.h"
#import "AdBrixJavascriptInterface.h"

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    WKWebViewConfiguration *config = [[WKWebViewConfiguration alloc] init];
    WKUserContentController *contentController = [[WKUserContentController alloc] init];
    [contentController addScriptMessageHandler:self name:@"adbrixBridge"];
    config.userContentController = contentController;

    self.webView = [[WKWebView alloc] initWithFrame:self.view.bounds configuration:config];
    [self.view addSubview:self.webView];

    NSURL *url = [NSURL URLWithString:@"https://web-sdk-dev-sec-dir.public.sre.dfinery.io/"];
    NSURLRequest *request = [NSURLRequest requestWithURL:url];
    [self.webView loadRequest:request];
    
    NSSet* nSet= [NSSet setWithArray:@[WKWebsiteDataTypeDiskCache, WKWebsiteDataTypeMemoryCache, WKWebsiteDataTypeCookies]];
    NSDate *nDate=[NSDate dateWithTimeIntervalSince1970:0];
    [WKWebsiteDataStore.defaultDataStore removeDataOfTypes:nSet modifiedSince:nDate completionHandler:^{}];

}

#pragma mark - WKScriptMessageHandler

- (void)userContentController:(WKUserContentController *)userContentController didReceiveScriptMessage:(WKScriptMessage *)message {
    if ([message.name isEqualToString:@"adbrixBridge"]) {
        [AdBrixJavascriptInterface event:message.body];
    }
}

@end

File: sample/ios/HybridTest/ViewController.swift
//
//  ViewController.swift
//  HybridTest
//
//  Created by Jimmy.강세훈 on 2023/03/14.
//

import UIKit
import WebKit

class ViewController: UIViewController {

    var webView: WKWebView!
    
    let interfaceName = "adbrixBridge"
    let webViewURL = "https://web-sdk-dev-sec-dir.public.sre.dfinery.io/"
    
    override func loadView() {
        let webConfiguration = WKWebViewConfiguration()
        let contentController = WKUserContentController()
        contentController.add(WeakContentController(delegate: self), name: interfaceName)
        webConfiguration.userContentController = contentController
        
        webView = WKWebView(frame: .zero, configuration: webConfiguration)
        
        view = webView
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        webViewInit()
    }
    
    private func webViewInit() {
        if let url = URL(string: webViewURL) {
            let request = URLRequest(url: url)
            webView.load(request)
        }
        
        //WebView cache 삭제
        let dataTypes = NSSet(array: [WKWebsiteDataTypeDiskCache, WKWebsiteDataTypeMemoryCache])
        let date = Date(timeIntervalSince1970: 0)
        WKWebsiteDataStore.default().removeData(ofTypes: dataTypes as! Set<String>, modifiedSince: date, completionHandler: {})
    }
}

extension ViewController: WKScriptMessageHandler {
    
    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        if message.name == interfaceName {
            AdbrixJavascriptInterface.event(message: message.body)
        }
    }
    
}

//Memory leak을 해결하기 위한 클래스
class WeakContentController: NSObject, WKScriptMessageHandler {
    weak var delegate: WKScriptMessageHandler?
    
    init(delegate: WKScriptMessageHandler) {
        self.delegate = delegate
        super.init()
    }
    
    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        self.delegate?.userContentController(userContentController, didReceive: message)
    }
}
